### DISCUSSION TITLE
Effect of gcc -g when compiling in different directories

### DISCUSSION THREAD

**Discussion by AlexanderLanin** [ccache/ccache#748] [General]
Sorry I'm abusing my support issue to document what I understood so far. Maybe it can be used to improve documentation.

When compiling with -g (gcc, clang) debug info will be added to the object file.
It can be observed by looking at `objdump -g your_object_file.o | grep -e DW_AT_name -e DW_AT_comp_dir | head`.
This debug info contains:
* a relative path from your build directory to the source file in question
* the build directory.

Therefore object files cannot be shared between different source or different build directories when using -g.

There are three ways to deal with this:
1) ignore the problem
2) unify paths before compilation so ccache can share cache
3) ccache could fix paths after extracting from cache
4) debuggers should just use executable file location and search from there

How to do that?
1) set `hash_dir` to false. However you are likely to run into surprises due to unexpected dependency to other source code locations. So your debugger could point you to the wrong code.

2) use compiler option `-fdebug-prefix-map={$build_dir}=.` to drop the build directory from the object file.
Problems:
* source directory is still relative. It seems you have to compile within your source directory so that it will end up with "../" which is always the same then. (ToDo: what would happen when compiling next to the source directory? Will ccache correctly not share the object file? It contains `../source_dir/`).
* debugger will not find source code.
  Workarounds include but are not limited to: start debugging from build directory or adding the build directory to paths where debugger will look for source files. https://chromium.googlesource.com/chromium/src/+/master/tools/gdb/gdbinit#82
  This seems gdb specific.

3) ccache would need to be able to read some formats, e.g. DWARF 4. Convert it. Save it. Unless all major compilers use a reasonably similar DWARF version and there is a convinient to use DWARF library out there this seems unfeasable.

4) It's been addressed ~2 years ago, but no progress: http://dwarfstd.org/ShowIssue.php?issue=171130.2

**Comment by jrosdahl** [comment]
> Maybe it can be used to improve documentation.

That would of course be welcome.
